File: src/main/java/io/javalin/translator/json/Jackson.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin.translator.json;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.javalin.HaltException;

import com.fasterxml.jackson.databind.ObjectMapper;

public class Jackson {

    private static Logger log = LoggerFactory.getLogger(Jackson.class);

    private static ObjectMapper objectMapper;

    public static String toJson(Object object) {
        if (objectMapper == null) {
            objectMapper = new com.fasterxml.jackson.databind.ObjectMapper();
        }
        try {
            return objectMapper.writeValueAsString(object);
        } catch (Exception e) {
            String message = "Failed to write object as JSON";
            log.warn(message, e);
            throw new HaltException(500, message);
        }
    }

    public static <T> T toObject(String json, Class<T> clazz) {
        if (objectMapper == null) {
            objectMapper = new com.fasterxml.jackson.databind.ObjectMapper();
        }
        try {
            return objectMapper.readValue(json, clazz);
        } catch (Exception e) {
            String message = "Failed to convert JSON to " + clazz.getName();
            log.warn(message, e);
            throw new HaltException(500, message);
        }
    }
}

File: src/main/java/io/javalin/translator/json/Jackson.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.translator.json

import com.fasterxml.jackson.databind.Module
import com.fasterxml.jackson.databind.ObjectMapper

object JavalinJacksonPlugin {

    private var objectMapper: ObjectMapper? = null

    private fun createObjectMapper(): ObjectMapper = try {
        val className = "com.fasterxml.jackson.module.kotlin.KotlinModule";
        ObjectMapper().registerModule(Class.forName(className)
                .getConstructor()
                .newInstance() as Module)
    } catch (e: ClassNotFoundException) {
        ObjectMapper()
    }

    @JvmStatic
    fun configure(staticObjectMapper: ObjectMapper) {
        objectMapper = staticObjectMapper
    }

    fun toJson(`object`: Any): String {
        objectMapper = objectMapper ?: createObjectMapper()
        return objectMapper!!.writeValueAsString(`object`)
    }

    fun <T> toObject(json: String, clazz: Class<T>): T {
        objectMapper = objectMapper ?: createObjectMapper()
        return objectMapper!!.readValue(json, clazz)
    }

}
